Rewrite of bookmark-it, fixing issue 182 and some other problems

wbamberg opened the pull request:
Here's a PR to fix https://github.com/mdn/webextensions-examples/issues/182. It's basically a rewrite, because it seems to me that storing `currentTab` and `currentBookmark` as global variables makes it easy for the add-on to get out of sync with reality. Also, the current add-on didn't handle changes to the bookmark store that happen outside the add-on itself (e.g. through the built-in bookmarking UI).

There's still a bug: bookmarks.search expects the `url` option to be in the `url` format which throws an error if the URL is like "about:debugging" [[1]](http://searchfox.org/mozilla-central/source/browser/components/extensions/schemas/bookmarks.json#258), [[2]](http://searchfox.org/mozilla-central/source/toolkit/components/extensions/Schemas.jsm#511). This seems like a WebExtensions bug, tbh, since it is possible to bookmark URLs like this, so it should be possible to search for those bookmarks too.

@kumar303 , do you think you could review this for me please?

kumar303 commented on bookmark-it/scripts/handle-button-click.js:
```diff
@@ -0,0 +1,31 @@
+
+  function updateBookmark(bookmarkArray) {
+    if (bookmarkArray.length === 0) {
+      browser.bookmarks.create({
```
Since you're using a `catch` function down below, the returned promise should be returned to the chain:
````js
return browser.bookmarks.create({...});
````

kumar303 commented on bookmark-it/scripts/handle-button-click.js:
```diff
@@ -0,0 +1,31 @@
+        url: tabInfo.url
+      });
+    } else {
+      browser.bookmarks.remove(bookmarkArray[0].id);
```
same here:
````js
return browser.bookmarks.remove(bookmarkArray[0].id);
````

kumar303 commented on bookmark-it/scripts/update-icon.js:
```diff
@@ -0,0 +1,91 @@
+*/
+function updateIcon(bookmarkArray) {
+  if (bookmarkArray.length === 0) {
+    browser.browserAction.setIcon({
```
Similarly, this promise should be returned so your call below can catch errors:
````js
return browser.browserAction.setIcon({...});
````

kumar303 commented on bookmark-it/scripts/update-icon.js:
```diff
@@ -0,0 +1,91 @@
+      path: notBookmarked
+    });
+  } else {
+    browser.browserAction.setIcon({
```
and this one too:
````js
return browser.browserAction.setIcon({...});
````

wbamberg commented:
@kumar303 , sorry to take so long to get to this. I don't think {currentWindow: true} does help: the main problem in the old version was that it wasn't listening for window switching. But in the end I revised the fix to be much smaller in scope.

I've added:
* listening to windows.onFocusChanged so it works with multiple windows
* ignoring protocols that will raise an error in bookmarks.search, and logging the ignoring
* listening for bookmarks being created and removed

kumar303 commented on bookmark-it/background.js:
```diff
@@ -42,28 +36,46 @@ browser.browserAction.onClicked.addListener(toggleBookmark);
+    var supportedProtocols = ["https:", "http:", "ftp:", "file:"];
+    var url = document.createElement('a');
+    url.href = urlString;
+    return supportedProtocols.indexOf(url.protocol) != -1;;
```
This line has an extra semi colon

wbamberg commented:
Thanks @kumar303 !
